### PR TITLE
Add support mCaptcha as captcha provider

### PR DESCRIPTION
**Pull request by Gusted**
- There's a deep motivation for this pull request that requires some explanation(feel free to skip reading this). In recent events, it was *once again* [pointed out](https://chaos.social/@dentangle/108652266792247321) that Codeberg is using a captcha that isn't accessible to blind and visually impaired people, currently Gitea doesn't offer a captcha that is accessible to those kind of people. After a [public meeting](https://codeberg.org/Codeberg/Community/issues/479#issuecomment-581498) was hold by Codeberg to discuss the issue and look into solutions, I came to a conclusion that [mCaptcha](https://mcaptcha.org/) is a good solution to solve Codeberg's accessibility problem. So therefor this PR.
- This PR adds a fourth captcha type, mCaptcha. It uses a [small library](https://codeberg.org/Gusted/mCaptcha) to interact with the mCaptcha instance. The javascript code uses the [official glue code (vanilla-variant)](https://github.com/mCaptcha/glue) provided by mCaptcha themselves. The remaining code is basically boilerplate that also exists for the other captchas.

### CODE REVIEW

File: web_src/js/features/mcaptcha.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    return;
+  }
+
+  const {default: mCaptcha} = await import(/* webpackChunkName: "mcaptcha-vanilla-glue" */'@mcaptcha/vanilla-glue');
+  mCaptcha.INPUT_NAME = 'm-captcha-response';
+  const siteKey = siteKeyEl.getAttribute('data-sitekey');
+
+  mCaptcha.default({
+    siteKey: {
+      instanceUrl: new URL('http://localhost:7000'),
```
Comment (6543) [inline web_src/js/features/mcaptcha.js]: hardcoded instance?

Comment (Gusted) [reply web_src/js/features/mcaptcha.js]: My bad. https://github.com/go-gitea/gitea/pull/20458/commits/5015c38cdd5cc7aaef6073bcefc6542d6cd99aa7

File: web_src/js/features/mcaptcha.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+export async function initMcaptcha() {
+  const mCaptchaEl = document.querySelector('.m-captcha');
+  if (!mCaptchaEl) {
+    return;
+  }
```
Comment (silverwind) [inline web_src/js/features/mcaptcha.js]:
```suggestion
  if (!mCaptchaEl) return;
```

Personal preference

File: package.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
   },
   "dependencies": {
     "@claviska/jquery-minicolors": "2.3.6",
+    "@mcaptcha/vanilla-glue": "0.1.0-alpha-1",
```
Comment (silverwind) [inline package.json]: alpha version seems not very promising, but I guess we are the first user of this according to download counts on https://www.npmjs.com/package/@mcaptcha/vanilla-glue.

Comment (Gusted) [reply package.json]: @realaravinth mentioned that the glue code is already "stable". So it shouldn't be anything major to bump the version to a stable 1.0 version.

Comment (realaravinth) [reply package.json]: The glue library has tests but I would like to wait for some deployment and see if any in-the-wild bugs are uncovered before I bump it to 1.0.

There won't be any changes in the library, it doesnt do much and what it does is simple enough to not require major changes.

At any rate, I will be available to handle support requests that are related to mCaptcha :)

**Comment by zeripath** [comment]
I don't understand though...  I thought the point of mCaptcha was that it was supposed to be a drop-in replacement for recaptcha/hcaptcha?

Can't you just put an mCaptcha url in as the recaptcha url?

**Comment by zeripath** [comment]
If not can we change our recaptcha and/or hcaptcha modules to make them a little more configurable so that they could handle mCaptcha or whatever other captcha solution comes in the future?

**Comment by realaravinth** [comment]
> I don't understand though... I thought the point of mCaptcha was that it was supposed to be a drop-in replacement for recaptcha/hcaptcha?
> 
> Can't you just put an mCaptcha url in as the recaptcha url?
> ### xCaptcha drop-in replacement
> 
> Already using a captcha solution? No worries, our APIs are compatible with reCaptcha and hCaptcha!

The website is out of date. I started out with the goal of building an identitical interface but hCaptcha and reCAPTCHA APIs have attributes that are irrelevant for mCaptcha. 

Current interface is similar but not identitical. Will update the website this week.

File: routers/web/auth/openid.go [hunk 4]
```diff
@@ -397,6 +401,12 @@ func RegisterOpenIDPost(ctx *context.Context) {
 				return
 			}
 			valid, err = hcaptcha.Verify(ctx, form.HcaptchaResponse)
+		case setting.MCaptcha:
+			if err := ctx.Req.ParseForm(); err != nil {
+				ctx.ServerError("", err)
+				return
+			}
+			valid, err = hcaptcha.Verify(ctx, form.HcaptchaResponse)
```
Comment (zeripath) [inline routers/web/auth/openid.go]: Shouldn't this be mcaptcha?

File: custom/conf/app.example.ini [hunk 5]
```diff
@@ -699,6 +699,8 @@ ROUTER = console
 ;; Type of captcha you want to use. Options: image, recaptcha, hcaptcha
```
Comment (fsologureng) [inline custom/conf/app.example.ini]:
```suggestion
;; Type of captcha you want to use. Options: image, recaptcha, hcaptcha, mcaptcha.
```

File: routers/web/auth/openid.go (hunk 4 above)
Comment (fsologureng) [inline routers/web/auth/openid.go]:
```suggestion
			valid, err = mcaptcha.Verify(ctx, form.McaptchaResponse)
```

File: package.json (hunk 3 above)
Comment (silverwind) [reply package.json]: Okay, please make sure to outline any future breaking changes before 1.0.0 in the GitHub release notes or a CHANGELOG file please.

File: templates/user/auth/signup_inner.tmpl [hunk 6]
```diff
@@ -54,6 +54,14 @@
 						<div class="h-captcha" data-sitekey="{{ .HcaptchaSitekey }}"></div>
 					</div>
 				{{end}}
+				{{if and .EnableCaptcha (eq .CaptchaType "mcaptcha")}}
+					<div class="inline field required df ac">
+						<label></label>
+						<div class="border-secondary" id="mcaptcha__widget-container" style="width: 304px; height: 78px"></div>
```
Comment (silverwind) [inline templates/user/auth/signup_inner.tmpl]: Can we avoid these specific width/height values?

**Review by silverwind — APPROVED** [review]
Looks reasonable except comment above, but I didn't test it.

File: templates/user/auth/signup_inner.tmpl (hunk 6 above)
Comment (Gusted) [reply templates/user/auth/signup_inner.tmpl]: mCaptcha's widget expect these dimensions(without these, it will not look centered)
image

**Comment by Gusted** [comment]
Blocking, due to fixed width, this doesn't work great on mobile.

File: templates/user/auth/signup_inner.tmpl (hunk 6 above)
Comment (silverwind) [reply templates/user/auth/signup_inner.tmpl]: Hmm okay. I have to say it looks a bit amateurish with non-vertically-centered content in the box and that big icon. Are you sure it's integrated correctly?

Comment (Gusted) [reply templates/user/auth/signup_inner.tmpl]: > I have to say it looks a bit amateurish with non-vertically-centered content in the box

That's without the explicit width/height, to show that it won't work. It actually looks like this:
image



> that big icon.

🤷🏽 it's what mCaptcha provides. 

> Are you sure it's integrated correctly?

Yes(*kind of*), but I'm working on mCaptcha to use better CSS, as currently it's broken for mobile.

Comment (silverwind) [reply templates/user/auth/signup_inner.tmpl]: > That's without the explicit width/height, to show that it won't work

That looks okay to me, at least from a vertical centering standpoint.

**Comment by Gusted** [comment]
Blocked by https://github.com/mCaptcha/glue/pull/12 & https://github.com/mCaptcha/mCaptcha/pull/35

**Comment by Gusted** [comment]
PRs are merged on mCaptcha. https://github.com/go-gitea/gitea/pull/20458/commits/c866103203575c193fdfb3095b55a8c3a83f2073 + https://github.com/go-gitea/gitea/pull/20458/commits/0789b3311d8ea1cdb768efa13a6ee050d76fbf1d fixes the issue on mobile.

File: package.json [hunk 7]
```diff
@@ -62,9 +63,5 @@
     "svgo": "2.8.0",
     "updates": "13.1.2"
   },
-  "browserslist": [
-    "defaults",
-    "not ie > 0",
-    "not ie_mob > 0"
-  ]
+  "browserslist": ["defaults", "not ie > 0", "not ie_mob > 0"]
```
Comment (silverwind) [inline package.json]: Unrelated

Comment (Gusted) [reply package.json]: https://github.com/go-gitea/gitea/pull/20458/commits/559cf0da2b997e3e5e86ef327f6e648fdb1d97ad IDE is pulling a funny over me by formatting this as a JSON file.

**Review by 6543 — APPROVED** [review]

**Comment by cryptoroland** [comment]
[](https://medium.com/@hCaptcha/how-hcaptcha-calculates-rewards-1195e6f18284)
